Improve handling of the integrate flag for models

DougBurke opened the pull request:
I'm coming back to review this and so the text below is not guaranteed to match up with the code behavior.

# Summary

Improve the handling of the integrate setting for models: it is now passed on for composite models, such as an expression like "mdl1 * (mdl2 - mdl3)", and XSPEC models no longer allow users to change the setting (any attempt will be ignored, which will allow scripts created by save_all using an older version of Sherpa to still work).

# Details

This is a rework of the last part of #947 and then merging in work done in #830. This is closely related to the regrid work but has been broken out of that (see #830, which depends on this PR). This PR depends on

- #1803

The ideas are that

- adding more tests to cover some of the issues we have uncovered
- allow the integrate to be passed "up" the model classes, so that composite models can have an integrate setting [1]
- special case the XSPEC models since we know that should send the data "integrated" arrays - that is low and high edges - **but** we also know there's a difference for additive and multiplicative models that maps well to `integrate=True` and `integrate=False` respectively when regridding data [2], [3]

[1] technically it makes to copy the integrate flag for something like an unary op model, but it's less clear what happens for binary op models where you may be combing models with `integrate=True` and `integrate=False`. We could set `integrate` to `None` in this case (or even record all the integrate settings in some way) but I am not sure it would help downstream users. I think the current code will fall over to a `True` setting for this case (see the paragraph starting "One interesting outcome" below).

[2] XSPEC models now have a fixed `integrate` setting, and attempting to change it does nothing. That is, you can say "mdl.integrate = False` for an additive model, but the field will stay set to `True`. I originally made this be an error, **but** that would break any use of `save_all` which contains a multiplicative XSPEC model, for no actual user benefit. So now we just do nothing. We could use the warning logger, or a warnings FutureWarning to create a message, but I am not sure many people actually change this setting, so all we'd then do would be make users of `save_all` see even more screen messages they would then learn to ignore.

[3] XSPEC convolution models are a special case - I believe the way that XSPEC works, you aren't able to apply a convolution model to anything but an additive model (that is, a model expression that "ends" with an additive component), which would imply XSPEC convolution models should be marked as `integrate=True`, so this is what we do

One interesting outcome of the current code, and something we could try to clean up, is that models tend now to default to `integrate=True` (thanks to `ArithmeticModel`) and that this can lead to things like "unary op model is sent something with no `integrate` setting, so we try not to set the `integrate` flag, but it magically gains it from the `ArithmeticModel` hierarchy".

It is hard to update the documentation to highlight this change, as there's no single `ui` routine (for example) we can document. Something to think about.

This code is going to conflict with all the existing XSPEC changes we have (e.g. #1396 https://github.com/sherpa/sherpa/pull/1609 https://github.com/sherpa/sherpa/pull/1615 https://github.com/sherpa/sherpa/pull/1630)

# Question

How useful is the `integrate` setting? In some ways I'd like to get rid of it, but that would be quite a large change for little obvious user benefit.

We could use the integrate setting to validate model expressions, in a similar manner to the way we do with model dimensionaluity, but

 - the dimensional check can be made when the model expression is being created, whereas the integrate setting would need to be checked whenever the composite model is evaluated
 - for XSPEC models we can come up with something (like you can multiply but not add multiplicative components) but is this true for the general case (and coming up with rules is more onerous than I'd like for a simple check)

# Trying to change the integrate setting

In #947 I had made code like

    mdl = XSapec()
    mdl.integrate = False

raise an error, because the integrate setting of this model (an additive model) is True. In this version it just ignores the user setting, so the following would hold after the above:

    assert mdl.integrate

Normally I'd be all for raising an error here, but it would make any script saved by `save_all` that contains an XSPEC multiplicative model to fail when evaluated because before this change the file would include a line `mdl.integrate = True` for these models  (because prior to this PR the multiplicative models did have integrate set to True), which would mean annoying changes for users that doesn't actually improve anything.  Hence the change to the setter just ignoring the user value.

# Example

## Generic

In CIAO 4.15 we have

```
sherpa-4.15.0> mdl1 = 2 + const1d.foo1

sherpa-4.15.0> mdl2 = 2 + scale1d.bar1

sherpa-4.15.0> foo1.integrate
True

sherpa-4.15.0> bar1.integrate
False

sherpa-4.15.0> mdl1.integrate
AttributeError: 'BinaryOpModel' object has no attribute 'integrate'

sherpa-4.15.0> mdl2.integrate
AttributeError: 'BinaryOpModel' object has no attribute 'integrate'

sherpa-4.15.0> tmp = -bar1

sherpa-4.15.0> tmp.integrate
AttributeError: 'UnaryOpModel' object has no attribute 'integrate'
```

With this PR we get

```
>>> from sherpa.astro.ui import *
>>> mdl1 = 2 + const1d.foo1
>>> mdl2 = 2 + scale1d.bar1
>>> (foo1.integrate, bar1.integrate)
(True, False)
>>> (mdl1.integrate, mdl2.integrate)
(True, True)
>>> tmp = -bar1
>>> tmp.integrate
False
```

Now, should `2 + bar1` have an integrate setting of `True`? Ideally not.

I note that when combining "actual" models we get the `integrate` setting combined as you might expect:

```
>>> foo1.integrate, bar1.integrate
(True, False)
>>> (foo1 + bar1).integrate
True
>>> (foo1 + foo1).integrate
True
>>> (bar1 + bar1).integrate
False
>>> (bar1 + foo1).integrate
True
```

[as a reminder, in CIAO 4.15 composite models do not have an `integrate` setting:

```
sherpa-4.15.0> (foo1 + bar1).integrate
AttributeError: 'BinaryOpModel' object has no attribute 'integrate'
```
]

## XSPEC

In CIAO 4.15 we have

```\sherpa-4.15.0> mdl = xsphabs.gal * xsapec.src

sherpa-4.15.0> (gal.integrate, src.integrate)
(True, True)

sherpa-4.15.0> mdl.integrate
AttributeError: 'BinaryOpModel' object has no attribute 'integrate'

sherpa-4.15.0> src.integrate = False

sherpa-4.15.0> (gal.integrate, src.integrate)
(True, False)
```

We now get

```
>>> mdl = xsphabs.gal * xsapec.src
>>> (gal.integrate, src.integrate)
(False, True)
>>> mdl.integrate
True
>>> (gal + gal).integrate     # NOTE: combing two False create a combined model with integrate=False
False
>>> src.integrate = False
>>> (gal.integrate, src.integrate)
(False, True)
>>> mdl.integrate
True
```

DougBurke committed:
xspec: integrate setting of XSPEC models is now fixed.

The integrate setting now matches the model type - so True for
Additive and Convolution, False for Mutiplicative - and can
not be changed.

However, to support old code that may have tried to change the
setting [*], we allow a call to try and change the integrate flag,
but do not actually change it. There is currently no warning about
this. Should there be (I can't make up my mind)?

The current meaning of the integrate flag is not well defined, so
this is a start to providing some meaning.

We could try and do things like "ensure convolution models are only
sent additive models (ie those with integrate=True), but that requires
more work (getting the integrate flag to be passed through the
*OpModel classes) and it's not entirely obvious it's worth it.

[*] Any save_all script created prior to this PR by Sherpa which
    contains any XSPEC multiplicative models will try to do this

DougBurke committed:
Unary/Binary models copy the integrate setting, if set

The integrate setting is copied from the components of a *OpModel
when it makes sense. For binary models, if both components are
False then the result is False, otherwise it's True. It is not
entirely clear, but for most cases it should default to True if
not explictily set (thanks to using multiple parent classes and
that one of them is ArithmeticModel) but it's not entirely clear
to me (e.g. ArithmeticFunctionModel should have no integrate
setting).

The idea is that, when combining models, we are going to want
to treat it like an additive model *unless* we know that both
sides are not additive.

hamogu commented:
I'm very opposed to just silently ignoring the setting. And, I've never used `save_all`. Maybe I should, I just never knew about it.
Anyway, a compromise could be to issue a warning "Setting the integrate setting on XSPEC models has no no effect. See URL HERE or release notes for details" or "This warning can be safely ignored, but don't do this from now on". Essentially, a Deprecation warning. Don't say "Hah. Error. Got you." Instead say "Setting integrate does not do anything, but we've always ignored that in the past. Now you at least get this warning and in the future it will be an error."

You could even make that an error for Additive models that don't have the `save_all` backwards compatibility problem.

hamogu commented:
Apart from the discussion on "should there be a message if I try to change it", the implementation seems uncontroversial. Since the setting is not used, cannot be changed and, for some models, also does not make any sense, I just want to ask: Do we really need an integrate setting at all? Or, if we need it, can be set is to `None` or the string "unused" so signify that it's in fact, not used.

DougBurke commented:
So, I would like to remove the `integrate` setting - which I believe would mainly be an issue for people using the Const*D models who should be using Scale*D instead (but our documentation points them to use the const version which we should address outside this PR).

However, I also want a way for us to identify whether a model behaves like the XSPEC additive versus multiplicative models, which I believe is important for us to get the regrid code working (and is the hidden driver behind this work). For that I think we need some way to say "hey, are you actually integrating the model or not" and I am co-opting the under-defined integrate setting to provide that. Perhaps they should be separate as this avoids some of this confusion (ie do we need to worry about existing code and the  integrate setting).

hamogu commented:
Either way works for me. I'd be OK with having as is and just adding an Exception/Warning for the cases where there is no backwards-compatibility issues (the Arithmetic models). We could then remove it two years from now.
Or, we could just add a warning for all and make `read_all` (is that the opposite of `save_all`?) just silence that warning. That way tou cna use te setting in any way way you like, I get the warning I want, and users of `save_all` are not unduly burdened with meaningless messages.

anetasie commented:
> So, I would like to remove the `integrate` setting - which I believe would mainly be an issue for people using the Const_D models who should be using Scale_D instead (but our documentation points them to use the const version which we should address outside this PR).
> 
> However, I also want a way for us to identify whether a model behaves like the XSPEC additive versus multiplicative models, which I believe is important for us to get the regrid code working (and is the hidden driver behind this work). For that I think we need some way to say "hey, are you actually integrating the model or not" and I am co-opting the under-defined integrate setting to provide that. Perhaps they should be separate as this avoids some of this confusion (ie do we need to worry about existing code and the integrate setting).**EDITED TO ADD** which doesn't really change this PR as such, but if we do decide to remove the integrate setting then I think we'd allow the integrate setting but warn users if they try to set it.

@DougBurke Do you want to remove the 'integrate' setting for all the models or only for XSPEC models?

DougBurke commented:
If we were to do it it would be for all models. However, as I have discussed above there are concerns I have over the regrid code - e.g. https://github.com/sherpa/sherpa/issues/829 - and so we may need some way of indicating how a model works when given lo/hi bins (is it integrating the data or not). XSPEC models have this by design (we have the XSAdditiveModel and XSMultiplicativeModel) but we need something for all models (whether it's a class or an attribute). I'm trying to get #830 to a place where we can discuss this but it's one of those annoying issues which doesn't really lend itself well to single PRs.

DougBurke commented:
I'm going to make this a draft PR for now, just to make sure it doesn't accidentally get merged.

anetasie commented:
If this is for all the models then it is quite a large change to our handling of the model evaluations. There is no distinction between multiplicative and additive models in Sherpa now, so each model could be combined in a general expression not dependent on the data type. If the integration is set by data then the multiplications or convolution expressions will be impacted by this change. As you described above this update requires more thoughts.

DougBurke commented:
@anetasie - it doesn't change how you combine models, or use convolution-style kernels, or how models work. There are two related but separate issues:

1. does changing the integrate flag make sense - I don't think people make use of the integrate setting: in part because we don't really advertise it but in part because I don't think it's that useful to be able to change and we have other ways we could handle if fir users who do want to switch the integrate setting
2. how does `regrid` work work for data evaluated on an integrated grid (which is #829)

anetasie commented:
@DougBurke I need to go back and check these other related issues.

DougBurke committed:
Tests: add testing of regrid/integrate for unop/binop models

Extend the recent support of regridding unop/binop models with
more tests.

There's also some minor code clean up in the tests.

DougBurke committed:
Tests: add XSPEC regrid/integration cases

Most tests actually fail, but are not marked as xfail as I want
to check we know wy they fail (during development of this work
there have been a number of changes to the system, and it's
hard to track what is really happening without explicit tests).

Note that the "combined" test - that is mdl1 * mdl2 - is tested
as both mul * add and add * mul since the regrid code currently
uses the first model when calculating the regrid method which
breaks here. I have also added a test where one of the models
is an ArithmeticConstant model, since they can have different
behavior to "normal" models.

There are also more "esoteric" tests of "deep" expressions,
like (2 * mdl1) / (3 + mdl2) - even if they don't make much
sense (and currently lead to recursion errors), and tests
related to unary and binomial combination of models.

Tests have also been added of combining Sherpa and XSPEC models
(one "additive" and one "multiplicative").

A pytest fixture - xsmodel - has been added to make it a little-bit
easier to create an XSPEC model (select the correct class and then
instatiate it).

DougBurke commented:
Rebased because of conflicts.

hamogu commented:
This is still marked as WIP, but it also has a request to review for me attached. So, just checking: Is this just open "for reference" as other PRS (like #1179) take charge of those changes or is there any actualy review here needed by me at this point?

DougBurke commented:
I don't think it needs a review at this time. I have a document I'm slowly working on trying to explain why I care about the integrate flag and what it has to do with the regrid code. It's slow progress, but probably worth SDS reviewing this before any PRs.

DougBurke committed:
XSPEC: allow models to be regridded

We drop the integrate setting so that the XSPEC models can be
regridded. Unfortunately the results are not particularly
encouraging, particularly for multiplicative models.

DougBurke commented:
Rebased as multiple things have changed, in particular #1617 has merged. Need to get back into this as it's not clear what is going on here...

DougBurke committed:
model: ArithmeticConstantModel now has integrate=False

ArithmeticConstantModels are considered to be "point" values as
they ignore the bin width (since they always return a value).

DougBurke committed:
Tests: add an example of a problematic regrid case

What happens when we regrid an "integrable" model (so a model that
needs lo/hi values) but has integrate=False behavior (particularly
if you then change the integrate setting)?

DougBurke committed:
Tests: add tests of issue #1802

DougBurke committed:
regrid: fix #1802

The problem was that when regrid was called on a BinaryOpModel
we just took the first component that had a regrid method, but
if this component was itself a BniaryOpModel - such as in

    mdl1 * mdl2 * mdl3

which turns into

    BinaryOpModel(*, BinaryOpModel(*, mdl1, mdl2), mdl3)

and so the regrid check will be on BinaryOpModel(*, mdl1, mdl2),
which has a regrid call but calls itself, leading to no-end of
recursion.

The solution is to explicitly handle BinaryOpModel cases and
hunt through the "full tree" for the first model that has a
regrid call (and is not a BinaryOpModel). The search is left-biased,
but there's no obvious benefit to any other approach.

Some alternatives include:

- explicitly storing the regrid "class" when creating the
  BinaryOpModel (probably in a _regrid field) so we can just
  identify it when calling regrid()

  this should work, but makes everyone pay for this support, when
  we can just do it when we actually need it

- we could use the ndim field for the model, and just pick
  RegriddableModel1D or RegriddableModel2D directly, which would
  be a lot easier, but it would not be extensible.

One concern is that this probably doesn't handle UnaryOpModel cases
well, but they are currently unsupported.

DougBurke committed:
regrid: support UnaryOpModel

Allow UnaryOpModel instances to be regridded.

DougBurke committed:
Tests: check UnaryOpModel regrid handling
